### ISSUE TITLE
Struggling with permissions issues for the AllUsers settings on Windows

### ISSUE THREAD

**Issue by moorepants** [conda/constructor#828]
### Checklist

- [X] I added a descriptive title
- [X] I searched open reports and couldn't find a duplicate

### What happened?

We build a constructor Windows installer annually and install for `AllUsers` in silent mode. This year many/most of the installed scripts in `<install-directory>/Scripts/` have permissions that prevent execution. For example, opening the Anaconda prompt and typing `Jupyter lab` gives `Error executing Jupyter command 'lab': [WinError 5] Access is denied`.   Similar errors are produced when trying to launch lab from the Anaconda Navigator. Running `spyder` from the prompt or Navigator works. We tried building the new installer with the constructor version (3.4.5) we used last year but still get the permission errors. We have many details of our trial and error in this issue: https://gitlab.ewi.tudelft.nl/bhmgerritsen/anaconda-dee-config/-/issues/64 but I'll summarize below.

Constructor setup:

- Channel: conda-forge (with anaconda-navigator, console_shortcut, menuinst, powershell_shortcut using `defaults::` notation and with conda channel priority set to "strict". The idea was to install everything form conda-forge except for these few anaconda specific packages.
- We install with silent mode like `tudelft-anaconda-v2024.01-Windows-x86_64.exe /InstallationType=AllUsers /RegisterPython=1 /S /D=C:\Programs\tudelft-anaconda` (but also have tried graphical installs)
- `construct.yaml` is here: https://gitlab.ewi.tudelft.nl/bhmgerritsen/anaconda-dee-config/-/blob/master/construct.yaml
- Our latest installer is here: https://gitlab.ewi.tudelft.nl/bhmgerritsen/anaconda-dee-config/-/releases

If we download the latest official Anaconda windows installer and run it in silent mode the permissions are correct. But our installer leaves many files with restricted permissions.

Changes from last year:

- The approach was have channel priority with `defaults` then `conda-forge` and install the `anaconda` package and then add various extra packages on top of anaconda from conda-forge.
- We set the `channel_priority=strict` this year and didn't last year.
- We hard pin the versions of all packages this year.
- Our constructor files from last year's release are here: https://gitlab.ewi.tudelft.nl/bhmgerritsen/anaconda-dee-config/-/tree/62d723fbbfabe495a0bb38c408b51a40dd1a8adc

We have run out of ideas as to what causes the permissions errors. Do you have any suggestions for what we may be doing wrong?

### Conda Info

```shell
NA
```

### Conda Config

```shell
NA
```

### Conda list

```shell
NA
```

**Comment by marcoesters** [comment]
I downloaded your 2024.01 and ran the GUI installation - everything looks good on my end. `jupyter lab` launches from the prompt without problems. I ran it on a Windows 11 machine where I'm the only user.

The fact that this also breaks with the old constructor makes me think that there is a specific interaction with your system. Does it maybe restrict what unsigned installers and/or binaries can do? The Anaconda and, since recently, the binaries  on Windows are all signed by Anaconda.

You could also check if installing the package from `defaults` instead of `conda-forge` helps.

**Comment by moorepants** [comment]
Thanks @marcoesters, good to know that it worked for you. We can look more closely at things in the environment we are installing to. I'll ask out IT if we need the signing and we're already working on checking the installs from defaults. I'll report back on our success/failure.

**Comment by moorepants** [comment]
@marcoesters did you do a "JustMe" installation? Because that does seem to work for us. It is the "AllUsers" option that has the borked permissions.

**Comment by moorepants** [comment]
One other important note I forgot from above, is that the installer we built last year works just fine with and AllUsers install on the computers this year. So that is odd.

**Comment by subsense33** [comment]
To reproduce the error, it is important to start the installer as the local SYSTEM account, not just as a local administrator. Our software deployment mechanism also installs software using the local SYSTEM user.

You can use [Microsoft psexec.exe](https://learn.microsoft.com/nl-nl/sysinternals/downloads/psexec) to start an elevated command prompt as the SYSTEM user.

1. Open an elevated Command Prompt.
2. Run the following command: `psexec.exe -s -i cmd`
3. A new Command Prompt window will open. To verify your current user, type `whoami`.

**Comment by moorepants** [comment]
> To reproduce the error, it is important to start the installer as the local SYSTEM account

Does that mean you have reproduced the error? This is exactly how we are installing using the SYSTEM account.

Oddly, the installer we build last year installs with correct permissions with the SYSTEM account but the new installer (built with either constructor 3.4.5 or the latest version) has incorrect permissions with the SYSTEM account.

**Comment by moorepants** [comment]
Ah, I didn't realize @subsense33 was my colleague here at TU Delft. He's just explaining what we've done. I assumed @subsense33 was another person commenting on this.

**Comment by moorepants** [comment]
I just installed the v2024.01 release from https://gitlab.ewi.tudelft.nl/bhmgerritsen/anaconda-dee-config/-/releases on a regular Windows PC. I selected the "All Users" option and to "Register Python". I elevated admin rights to run the installer and then afterwards checked running spyder, jupyter, python console, ipython both from command line and anaconda navigator app and everything worked correctly with no permissions issues. This is presumably what @marcoesters did.

So, this points towards some kind of issue installing with the local SYSTEM account. (except that last year's installer has no issue installing with the SYSTEM account)

**Comment by marcoesters** [comment]
Thanks @subsense33 for these instructions! I was able to reproduce the problem now, but I suspect this is not related to `constructor` or your installer, but to `jupyterlab`. Here is what I did to come to that conclusion:

* Open Anaconda Prompt
* Run `jupyter lab` --> `WinError 5`. I also ran `jupyter notebook` with the same issue. Other executable like `spyder` open without errors.
* Create an environment in a location writable to me: `conda create -p %USERPROFILE%\testenv`
* `conda activate %USERPROFILE%\testenv`
* `conda install -y jupyterlab -c conda-forge`
* Run `jupyter lab` --> `jupterlab` opens without error

My hypothesis is that `jupyterlab` tries to write files into its own install location, which is write-protected. This would also explain why rebuilding the installer with an older `constructor` version doesn't work either. Your installer from last year (I assume) contains an older version of `jupyterlab`, which may not exhibit this behavior.

**Comment by marcoesters** [comment]
Of course, executing `jupyterlab` from the new environment means that I have full permissions to that executable as well. However, the fact that I can execute all other binaries in the base environment tells me that `jupyter` tries to do some kind of write operation into a location that it's not allowed to do.

**Comment by marcoesters** [comment]
I also manually checked the permissions inside the `Scripts` folder and they are what I expect: as a user, I only have read and execute permissions. The `WinError` is also a python error, so it reaches and executes the python interpreter somehow. If this was a permission error of your directory, I'd expect the error to be emitted by Windows, not python.

**Comment by moorepants** [comment]
Thanks for the insight. The permission error also occurs when trying to run `ipython` from the anaconda prompt.

We could install an older version of IPython and Jupyter as a stopgap.

**Comment by moorepants** [comment]
Here is an issue on the IPython tracker that implies that IPython tries to create something in the user's home directory:

https://github.com/ipython/ipython/issues/12079

It is from 2020 though, so that doesn't align with our last year's installer working.

**Comment by moorepants** [comment]
Searching for "jupyter" and "access is denied" gives lots of hits, like: https://stackoverflow.com/questions/61747187/jupyter-lab-in-anaconda-not-working-error-executing-jupyter-command-lab-win

**Comment by moorepants** [comment]
It's worth noting that we see this:
![image](https://github.com/user-attachments/assets/91b98fe7-d7b1-43a5-82c4-b8798cd60376)

indicating that we don't even have read and execute permissions for all the binaries.

**Comment by marcoesters** [comment]
You are right, alembic and other binaries don't have the correct permissions. jupyter does though.

For `AllUsers` installations, the permissions are supposed to be fixed here: https://github.com/conda/constructor/blob/main/constructor/nsis/main.nsi.tmpl#L1240-L1253

You can monitor the installation and see that you will lose read access to the entire directory during the installation, but get it back shortly after. Maybe the permissions changes are not performed recursively? I will have to investigate what those commands do exactly, but that change is also two years old, so this cannot be the only factor.

Alembic doesn't have read access throughout the entire installation process, even before the permissions are fixed in the lines above.

You may be running into a few things at once here:

* jupyter has the correct permissions, so you may see the problem I described earlier. This isn't something `constructor` will be able to change, but maybe you can configure jupyter/ipython in a post-install script to write somewhere else.
* As for alembic et al., `constructor` may have to retroactively change the permissions recursively if this isn't done already. I will have to think about whether this comes with security risks, even though read and execute privileges should be relatively safe.

**Comment by moorepants** [comment]
> I will have to think about whether this comes with security risks, even though read and execute privileges should be relatively safe.

Our current work around is to give everything in `<install-dir>/Scripts` read/exec permissions in a post install script.

**Comment by subsense33** [comment]
> > I will have to think about whether this comes with security risks, even though read and execute privileges should be relatively safe.
> 
> Our current work around is to give everything in `<install-dir>/Scripts` read/exec permissions in a post install script.

Yesterday I changed the work-around script to reset the owner on `<install-dir>`. Not just `<install-dir>/Scripts`. Does no harm, just to be sure everything is set correctly.
The script resets the owner of the entire tree <install-dir> to the local group "Administrators", that is how it's supposed to be. (Like the Anaconda deployment of past years).

**Comment by marcoesters** [comment]
While that workaround probably works, the ownership of the files is not different from the Anaconda installer. I am using `jupyter-kernel.exe` as an example here. File permissions are listed using `Get-Acl | fl`.

Anaconda installer, installed as `SYSTEM`:

```Powershell
Path   : Microsoft.PowerShell.Core\FileSystem::C:\ProgramData\anaconda3\Scripts\jupyter-kernel.exe
Owner  : BUILTIN\Administrators
Group  : NT AUTHORITY\SYSTEM
Access : BUILTIN\Users Allow  ReadAndExecute, Synchronize
         NT AUTHORITY\SYSTEM Allow  FullControl
         BUILTIN\Administrators Allow  FullControl
Audit  :
Sddl   : O:BAG:SYD:AI(A;ID;0x1200a9;;;BU)(A;ID;FA;;;SY)(A;ID;FA;;;BA)
```

TU Delft installer, installed as `SYSTEM`:

```Powershell
Path   : Microsoft.PowerShell.Core\FileSystem::C:\ProgramData\tudelft-anaconda\Scripts\jupyter-kernel.exe
Owner  : BUILTIN\Administrators
Group  : NT AUTHORITY\SYSTEM
Access : NT AUTHORITY\SYSTEM Allow  FullControl
         BUILTIN\Administrators Allow  ReadAndExecute, Synchronize
Audit  :
Sddl   : O:BAG:SYD:PAI(A;;FA;;;SY)(A;;0x1200a9;;;BA)
```

So, the owner is the same, but for some reason, access isn't granted the same way. I'm wondering if this is because of differences between the two build systems (Anaconda vs. conda-forge).

I'd be really curious to see if the problem persists if you build the installer using only `defaults` packages.

**Comment by moorepants** [comment]
This was my attempt at building the new installer the same way we did last year (defaults as priority w/ conda forge packages on top):

https://github.com/moorepants/anaconda-dee-config/actions/runs/10185441100

We have the same permissions issue with it.

I can build an installer without any packages from conda forge, but I just have to go through an make a new `construct.yaml` that eliminates all packages from conda forge. Last year's approach included the `anaconda` package (as does the one I linked just above). Are you suggesting to build an installer only from defaults package but without using the `anaconda` package or to include it?

Note that I switched to the "conda forge only" approach because it has long been recommend to not mix conda forge packages with defaults packages by conda forge. Our users need packages only available on conda forge, so I went the "pure" conda forge route to ensure binary compatibility.

**Comment by marcoesters** [comment]
I suggest to not use `conda-forge` as a channel for this test, yes. I'd also reduce the number of packages since we know which packages are problematic: the packages that install alembic or jupyter-kernel. That just makes debugging faster.

If that results in the same problem, I'm out of ideas. The workflow looks okay. The only difference I see is that you use PowerShell instead of bash (as Miniforge and Anaconda do), but that can't be the reason, can it?

**Comment by moorepants** [comment]
> PowerShell instead of bash (as Miniforge and Anaconda do), but that can't be the reason, can it?

I can test a build using bash to call the constructor related commands.

**Comment by marcoesters** [comment]
> I built an installer without conda-forge packages in the base environment. You can download here:
> 
> https://github.com/moorepants/anaconda-dee-config/actions/runs/10242877472
> 
> Note that my install action fails with:
> 
> ```
> Error: Unable to locate executable file: C:\Users\runneradmin\miniconda3\condabin\conda.bat. Please verify either the file path exists or the file can be found within a directory specified by the PATH environment variable. Also verify the file has a valid extension for an executable file.
> ```
> 
> I'm not sure why. I did not include the `anaconda` package.

Probably because the installation has failed. I'd run the GUI with the artifact and/or create an environment with your specs to see if there is a package conflict.

**Comment by moorepants** [comment]
This is the new installer we build this year with Constructor 3.11.3:: https://github.com/moorepants/anaconda-dee-config/actions/runs/14701742124/artifacts/3019641164

**Comment by moorepants** [comment]
Looking at the conversation last year, I guess we still need to make an installer with a minimal set of packages and see if it also has the permissions issues.

One thing I just thought of. It might be that the last working installer (without the permissions issues) was built on one of our university Windows desktops and the ones with the permissions issues are the ones I build in the Github Actions CI. I will look into this.

**Comment by moorepants** [comment]
There was also this issue https://github.com/conda/constructor/issues/872 that is more recent that is also related to permissions. May be worth comparing this issue and that one.

**Comment by moorepants** [comment]
I see that @Dribbelswag did build the 2023 and 2024 versions on an identical computer (not GIthub CI) here: https://gitlab.ewi.tudelft.nl/bhmgerritsen/anaconda-dee-config/-/issues/64#note_2408814 The errors occur in the 2024+ versions whether we build locally or on CI.

**Comment by subsense33** [comment]
> I'll try to build a defaults only installer.

Great, just give me a nudge via Teams when I can test the installer. I'll report the results back here afterwards

**Comment by moorepants** [comment]
I built a defaults only installer here: https://github.com/moorepants/anaconda-dee-config/actions/runs/14863287877 It has most of the packages we had in the conda-forge build.

**Comment by Dribbelswag** [comment]
@marcoesters 
I have tested the "Default only installer" from @moorepants and I see no issues regarding NTFS/Owner permissions.
Jupyter Notebook starts correctly without an Access Denied.

**Comment by moorepants** [comment]
Note that I had to remove a fair number of the installed packages that were only available on conda-forge (diff: https://github.com/moorepants/anaconda-dee-config/pull/23/files) so I wouldn't know how to identify which package(s) relate to this permission issue. At least we now know that packages from defaults and packages from conda-forge do something different.

**Comment by marcoesters** [comment]
> Note that I had to remove a fair number of the installed packages that were only available on conda-forge

Building the installer with the same packages with `defaults` and `conda-forge` would eliminate that variable, but we can probably learn something from this. Do you see same difference in the file permissions as in [this comment](https://github.com/conda/constructor/issues/828#issuecomment-2265664158) from last year?

> At least we now know that packages from defaults and packages from conda-forge do something different.

That's what I suspect, but the solution could still be in `constructor`. We may need to adjust how the permissions are set.

**Comment by moorepants** [comment]
> Building the installer with the same packages with defaults and conda-forge would eliminate that variable

I can build a conda-forge only with the reduce set of packages.

> Do you see same difference in the file permissions as in https://github.com/conda/constructor/issues/828#issuecomment-2265664158 from last year?

@Dribbelswag can you run the `Get-Acl | fl` command and report the results?

**Comment by subsense33** [comment]
Non-working installation (Installed under the local SYSTEM account):

`PS C:\WINDOWS\system32> Get-Acl "C:\Programs\tudelft-anaconda\Scripts\jupyter-kernel.exe" | Format-List


Path   : Microsoft.PowerShell.Core\FileSystem::C:\Programs\tudelft-anaconda\Scripts\jupyter-kernel.exe
Owner  : NT AUTHORITY\SYSTEM
Group  : NT AUTHORITY\SYSTEM
Access : OWNER RIGHTS Allow  ReadPermissions
         NT AUTHORITY\SYSTEM Allow  FullControl
         NT SERVICE\Winmgmt Allow  FullControl
Audit  :
Sddl   : O:SYG:SYD:PAI(A;;RC;;;OW)(A;;FA;;;SY)(A;;FA;;;S-1-5-80-3750560858-172214265-3889451188-1914796615-4100997547)



PS C:\WINDOWS\system32> (Get-Acl "C:\Programs\tudelft-anaconda\Scripts\jupyter-kernel.exe").Access


FileSystemRights  : ReadPermissions
AccessControlType : Allow
IdentityReference : OWNER RIGHTS
IsInherited       : False
InheritanceFlags  : None
PropagationFlags  : None

FileSystemRights  : FullControl
AccessControlType : Allow
IdentityReference : NT AUTHORITY\SYSTEM
IsInherited       : False
InheritanceFlags  : None
PropagationFlags  : None

FileSystemRights  : FullControl
AccessControlType : Allow
IdentityReference : NT SERVICE\Winmgmt
IsInherited       : False
InheritanceFlags  : None
PropagationFlags  : None`

Working installation from @moorepants 

`PS C:\WINDOWS\system32> Get-Acl "C:\Programs\tudelft-anaconda\Scripts\jupyter-kernel.exe" | Format-List


Path   : Microsoft.PowerShell.Core\FileSystem::C:\Programs\tudelft-anaconda\Scripts\jupyter-kernel.exe
Owner  : NT AUTHORITY\SYSTEM
Group  : NT AUTHORITY\SYSTEM
Access : DASTUD\Domain Users Allow  ReadAndExecute, Synchronize
         BUILTIN\Users Allow  ReadAndExecute, Synchronize
         BUILTIN\Administrators Allow  FullControl
         NT AUTHORITY\SYSTEM Allow  FullControl
Audit  :
Sddl   : O:SYG:SYD:AI(A;ID;0x1200a9;;;DU)(A;ID;0x1200a9;;;BU)(A;ID;FA;;;BA)(A;ID;FA;;;SY)



PS C:\WINDOWS\system32> (Get-Acl "C:\Programs\tudelft-anaconda\Scripts\jupyter-kernel.exe").Access


FileSystemRights  : ReadAndExecute, Synchronize
AccessControlType : Allow
IdentityReference : DASTUD\Domain Users
IsInherited       : True
InheritanceFlags  : None
PropagationFlags  : None

FileSystemRights  : ReadAndExecute, Synchronize
AccessControlType : Allow
IdentityReference : BUILTIN\Users
IsInherited       : True
InheritanceFlags  : None
PropagationFlags  : None

FileSystemRights  : FullControl
AccessControlType : Allow
IdentityReference : BUILTIN\Administrators
IsInherited       : True
InheritanceFlags  : None
PropagationFlags  : None

FileSystemRights  : FullControl
AccessControlType : Allow
IdentityReference : NT AUTHORITY\SYSTEM
IsInherited       : True
InheritanceFlags  : None
PropagationFlags  : None



PS C:\WINDOWS\system32>`

**Comment by moorepants** [comment]
Here is an installer with conda-forge packages that match the packages in https://github.com/conda/constructor/issues/828#issuecomment-2858272572:

https://github.com/moorepants/anaconda-dee-config/actions/runs/14898665479

Also the associated PR: https://github.com/moorepants/anaconda-dee-config/pull/24

@Dribbelswag can you install this one and check the permissions like the other (this ensures we are comparing installers that are identical other than the source channel from the packages).

Thanks.

**Comment by moorepants** [comment]
Just for future reference this is the `construct.yaml` file for the installer:

```yaml
name: tudelft-conda
version: v2025.01.dev
company: Delft University of Technology
installer_type: all
channels:
  - conda-forge  # change to "defaults" to get correct permissions
specs:
  - astropy
  - autopep8
  - black
  - bokeh
  - bzip2
  - cloudpickle
  - colorama
  - colorful
  - conda
  - configparser
  - curl
  - cython
  - dash
  - dask
  - dataclasses
  - datashader
  - dill
  - flask
  - gmpy2
  - graphviz
  - h5py
  - holoviews
  - intervals
  - ipykernel
  - ipympl
  - ipython
  - ipywidgets
  - jedi
  - jupyter
  - jupyterlab
  - jupytext
  - markdown
  - matplotlib
  - mccabe
  - menuinst
  - more-itertools
  - mpmath
  - mypy
  - nb_conda_kernels
  - nbclassic
  - nbclient
  - nbconvert
  - nbconvert-core
  - nbconvert-pandoc
  - nbformat
  - nbgrader
  - networkx
  - nose
  - notebook
  - numba
  - numexpr
  - numpy
  - packaging
  - pandas
  - pillow
  - pip
  - pyflakes
  - pygments
  - pygraphviz
  - pylint
  - pyqt
  - pytables
  - pytest
  - python
  - pyyaml
  - qtconsole
  - requests
  - rope
  - ruamel_yaml
  - scikit-image
  - scikit-learn
  - scipy
  - scrapy
  - seaborn
  - setuptools
  - sphinx
  - spyder
  - spyder-kernels
  - sqlalchemy
  - sqlite
  - statsmodels
  - sympy
  - toolz
  - traitlets
  - tzdata
  - urllib3
  - wheel
  - xarray
build_outputs:
  - info.json
  - pkgs_list
```

Switching the channel `conda-forge` to `defaults` is the only change. The defaults one has correct permissions and the conda-forge one does not.

**Comment by marcoesters** [comment]
The SDDL strings show that the files from `conda-forge` have the "protected" flag set (`O:SYG:SYD:PAI`) as opposed to `defaults` (`O:SYG:SYD:AI`). I don't know where it comes from or how this can be unset with NSIS, but I'll debug this more either tomorrow or next week.

**Comment by marcoesters** [comment]
I figured out why this is happening, but I am not sure about how to solve this yet. The cliff notes are:

* The only files that are affected are entry point exe files. Packages on `defaults` provide these as part of the package whereas for `conda-forge` packages, they are created on-the-fly by `conda`.
* These on-the-fly endpoints are hard links to `Scripts\conda.exe`.
* During the installation, this file is provided by `_conda.exe` (via [conda-standalone](https://github.com/conda/conda-standalone)). This file is created by PyInstaller, so it extracts itself into a temporary directory.
* When creating a hard link from this temporary directory into the installation directory, Windows adds the `P` flag to the access control flags. This happens with any user account, not just `SYSTEM`.

I'm thinking through possible solutions because that will affect all `noarch` packages that have an entry point inside `Scripts`.

**Comment by moorepants** [comment]
Nice work! Thanks for digging into this.

> When creating a hard link from this temporary directory into the installation directory, Windows adds the P flag to the access control flags. This happens with any user account, not just SYSTEM.

For this note, we have not seen permissions issues when installing in normal user accounts. Is it that the P flag is present but has not adverse effects for non-SYSTEM accounts?

**Comment by marcoesters** [comment]
That's because the permissions are already set correctly. The installer relies on inheritance to change the permissions for all-users installations and the `P` flag prohibits  that. The permissions are also incorrect when installing as a local administrator, but we probably were able to read the file because we are in the local administrator group.

So, I understand what needs to be done, but I still need to find the best/most efficient way to get there.

**Comment by moorepants** [comment]
Ok, clear. Thanks for looking into this.

**Comment by marcoesters** [comment]
#991 should solve this.

**Comment by moorepants** [comment]
Great news and very nice work!

**Comment by subsense33** [comment]
> [@subsense33](https://github.com/subsense33) I built an installer based on the [#991](https://github.com/conda/constructor/pull/991) changes, which can be downloaded here: https://github.com/moorepants/anaconda-dee-config/actions/runs/15697333735 . Can you check if the permissions are fixed on our TU Delft computers? Thanks.

@moorepants This looks very good.
I installed this release **without** the ResetOwner.ps1 workaround (from Software Center on Win10)
I can start Spyder, Jupyter Notebook and Lab without any permission issues.

I think you guys nailed it, well done!
